### PR TITLE
Feature/check address btc 183

### PR DESCRIPTION
**Pull request by blackyblack**

### CODE REVIEW

File: include/veriblock/entities/address_entity.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#define ALT_INTEGRATION_ADDRESS_ENTITY_HPP
+
+#include <cstdint>
+#include <string>
+
+#include "veriblock/consts.hpp"
+
+namespace VeriBlock {
+
+struct AddressEntity {
```
Comment (Warchant) [inline include/veriblock/entities/address_entity.hpp]: why renamed to AddressEntity?

and why removed existing serialization/deserialization functions?

File: src/address.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  }
+  return false;
+}
+
+static bool isBase59String(const std::string& input) {
+  try {
+    auto decoded = DecodeBase59(input);
+    (void)decoded;
+    return true;
+  } catch (std::invalid_argument e) {
```
Comment (Warchant) [inline src/address.cpp]:
```suggestion
  } catch (const std::invalid_argument& e) {
```

File: test/entities/address_test.cpp [hunk 3]
```diff
@@ -12,32 +12,94 @@ static auto ADDRESS_BYTES =
...
+}
+
+TEST(Address, DerivedFromPublicKey) {
+  auto publicKey =
+      "3056301006072a8648ce3d020106052b8104000a03420004cb427e41a0114874080"
+      "a4b1e2ab7920e22cd2d188c87140defa447ee5fc44bb848e1c0db5ef206de2e7002"
+      "f6c86952be4823a4c08e65e4cdbeb904a8b95763aa"_unhex;
+  std::string addressString = "VFFDWUMLJwLRuNzH4NX8Rm32E59n6d";
+  AddressEntity address = addressFromString(addressString);
+  EXPECT_EQ(addressIsDerivedFromPublicKey(address, publicKey),
```
Comment (Warchant) [inline test/entities/address_test.cpp]:
```suggestion
  EXPECT_TRUE(addressIsDerivedFromPublicKey(address, publicKey));
```

File: test/entities/address_test.cpp [hunk 4]
```diff
@@ -12,32 +12,94 @@ static auto ADDRESS_BYTES =
...
+}
+
+TEST(Address, NotDerivedFromPublicKey) {
+  auto publicKey =
+      "3056301006072a8648ce3d020106052b8104000a03420004cb427e41a0114874080"
+      "a4b1e2ab7920e22cd2d188c87140defa447ee5fc44bb848e1c0db5ef206de2e7002"
+      "f6c86952be4823a4c08e65e4cdbeb904a8b95763aa"_unhex;
+  std::string addressString = "V23Cuyc34u5rdk9psJ86aFcwhB1md0";
+  AddressEntity address = addressFromString(addressString);
+  EXPECT_EQ(addressIsDerivedFromPublicKey(address, publicKey), false);
```
Comment (Warchant) [inline test/entities/address_test.cpp]:
```suggestion
  EXPECT_FALSE(addressIsDerivedFromPublicKey(address, publicKey));
```

File: include/veriblock/entities/address_entity.hpp (hunk 1 above)
Comment (blackyblack) [reply include/veriblock/entities/address_entity.hpp]: I want to separate data structures and associated parse/serialize/validate functions. We either move everything to address.hpp or move everything out of there. I've chosen the latter.

File: src/address.cpp (hunk 2 above)
Comment (blackyblack) [reply src/address.cpp]: Does not compile on my side.

File: src/address.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  auto data =
+      std::string{"V"} + keyHashEncoded.substr(0, MULTISIG_ADDRESS_DATA_END);
+  auto checksum = calculateChecksum(data, false);
+  return AddressEntity(AddressType::STANDARD, data + checksum);
+}
+
+bool addressIsDerivedFromPublicKey(const AddressEntity& address,
+                                   Slice<const uint8_t> publicKey) {
+  auto expectedAddress = addressFromPublicKey(publicKey);
+  if (!(address == expectedAddress)) {
```
Comment (Warchant) [inline src/address.cpp]:
```suggestion
  if (address != expectedAddress) {
```

File: src/address.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    return true;
+  } catch (std::invalid_argument e) {
+    // do not throw - return status instead
+  }
+  return false;
+}
+
+static std::string calculateChecksum(const std::string& data, bool multisig) {
+  std::vector<uint8_t> dataBytes = toBytes(data);
+  auto hash = sha256(dataBytes);
```
Comment (Warchant) [inline src/address.cpp]: no need to copy string to vector and then calculate sha256 from vector. you can calculate sha256 directly from string

**Pull request merged**